cmd/vic-machine/ops-user/rbac.go

```diff
@@ -0,0 +1,286 @@
+	Resources []Resource
+}
+
+type PermissionList []types.Permission
+
+var rolePrefix = "vch-"
+
+var RoleVCenter = types.AuthorizationRole{
+	Name: "vcenter",
+	Privilege: []string{
+		"Datastore.Config",
+	},
+}
+
+var RoleDataCenter = types.AuthorizationRole{
+	Name: "datacenter",
+	Privilege: []string{
+		"Datastore.Config",
+		"Datastore.FileManagement",
+	},
+}
+
+var RoleDataStore = types.AuthorizationRole{
+	Name: "datastore",
+	Privilege: []string{
+		"Datastore.AllocateSpace",
+		"Datastore.Browse",
+		"Datastore.Config",
+		"Datastore.DeleteFile",
+		"Datastore.FileManagement",
+		"Host.Config.SystemManagement",
+	},
+}
+
+var RoleNetwork = types.AuthorizationRole{
+	Name: "network",
+	Privilege: []string{
+		"Network.Assign",
+	},
+}
+
+var RoleEndpoint = types.AuthorizationRole{
+	Name: "endpoint",
+	Privilege: []string{
+		"DVPortgroup.Create",
+		"DVPortgroup.Delete",
+		"DVPortgroup.Modify",
+		"DVPortgroup.PolicyOp",
+		"DVPortgroup.ScopeOp",
+		"VApp.AssignVM",
+		"VirtualMachine.Config.AddNewDisk",
+		"VirtualMachine.Config.AdvancedConfig",
+		"VirtualMachine.Config.EditDevice",
+		"VirtualMachine.Config.RemoveDisk",
+		"VirtualMachine.GuestOperations.Execute",
+		"VirtualMachine.Interact.DeviceConnection",
+		"VirtualMachine.Interact.PowerOff",
+		"VirtualMachine.Interact.PowerOn",
+		"VirtualMachine.Inventory.Create",
+		"VirtualMachine.Inventory.Delete",
+		"VirtualMachine.Inventory.Register",
+		"VirtualMachine.Inventory.Unregister",
+	},
+}
+
+var RBACConf = RBACConfig{
+	Resources: []Resource{
+		{
+			Type:      VCenter,
+			Propagate: false,
+			Role:      RoleVCenter,
+		},
+		{
+			Type:      Datacenter,
+			Propagate: false,
+			Role:      RoleDataCenter,
+		},
+		{
+			Type:      Cluster,
+			Propagate: true,
+			Role:      RoleDataStore,
+		},
+		{
+			Type:      VSANDatastore,
+			Propagate: false,
+			Role:      RoleDataStore,
+		},
+		{
+			Type:      DatastoreFolder,
+			Propagate: true,
+			Role:      RoleDataStore,
+		},
+		{
+			Type:      Network,
+			Propagate: false,
+			Role:      RoleNetwork,
+		},
+		{
+			Type:      Endpoint,
+			Propagate: true,
+			Role:      RoleEndpoint,
+		},
+	},
+}
+
+func NewAuthzManager(ctx context.Context, session *session.Session, principal string) *AuthzManager {
+	authManager := object.NewAuthorizationManager(session.Vim25())
+	mgr := &AuthzManager{
+		session:      session,
+		authzManager: authManager,
+		principal:    principal,
+	}
+	return mgr
+}
+
+func (am *AuthzManager) CreateRoles(ctx context.Context) (int, error) {
+	return am.createOrRepairRoles(ctx, getTargetRoles())
+}
+
+func (am *AuthzManager) DeleteRoles(ctx context.Context) (int, error) {
+	return am.deleteRoles(ctx, getTargetRoles())
+}
+
+func (am *AuthzManager) RoleList(ctx context.Context) (object.AuthorizationRoleList, error) {
+	return am.getRoleList(ctx)
+}
+
+func (am *AuthzManager) createOrRepairRoles(ctx context.Context, targetRoles []types.AuthorizationRole) (int, error) {
```

Comment:
We may need to be able to report the missing privileges/roles/permissions in the case where the user has _not_ checked the `Grant Permissions` box. You should consider splitting assembly/identification of missing permissions from the modification - maybe that's a follow up PR after function's complete but will likely reduce duplication.

Reply:
I am thinking of using a different approach for verification of pre-existing permissions. I would log in as the ops-user and use the vmomi API that checks if the user has the permission to do an operation on a resource. I still have to investigate the details of that would work but I doubt there will be much code sharing at the Role level.